adapters/ix/ix.go

```diff
@@ -1,267 +1,189 @@
-			// ext is DFP div ID and KV pairs if avail
-			//indexReq.Imp[i].Ext = json.RawMessage("{}")
-
-			// Any objects pointed to by indexReq *must not be mutated*, or we will get race conditions.
-			siteCopy := *indexReq.Site
-			siteCopy.Publisher = &openrtb.Publisher{ID: params.SiteID}
-			indexReq.Site = &siteCopy
+	// Multi-size banner imps are split into single-size requests.
+	// The first size imp requests are added to the first slice.
+	// Additional size requests are added to the second slice and are merged with the first at the end.
+	// Preallocate the max possible size to avoid reallocating arrays.
+	requests := make([]*adapters.RequestData, 0, maxRequests)
+	multiSizeRequests := make([]*adapters.RequestData, 0, maxRequests-nImp)
+	errs := make([]error, 0, 1)
 
-			// spec also asks for publisher id if set
-			// ext object on request for prefetch
-			j, _ := json.Marshal(indexReq)
+	headers := http.Header{
+		"Content-Type": {"application/json;charset=utf-8"},
+		"Accept":       {"application/json"}}
 
-			request := callOneObject{requestJSON: *bytes.NewBuffer(j), width: format.W, height: format.H}
-
-			// prioritize slots over sizes
-			if sizeIndex == 0 {
-				prioritizedRequests = append(prioritizedRequests, request)
-			} else {
-				requests = append(requests, request)
-			}
-		}
-	}
-
-	// cap the number of requests to requestLimit
-	requests = append(prioritizedRequests, requests...)
-	if len(requests) > requestLimit {
-		requests = requests[:requestLimit]
-	}
-
-	if len(requests) == 0 {
-		return nil, &errortypes.BadInput{
-			Message: "Invalid ad unit/imp/size",
+	imps := request.Imp
+	for iImp := range imps {
+		request.Imp = imps[iImp : iImp+1]
+		if request.Site != nil {
+			setSitePublisherId(request, iImp)
 		}
-	}
 
-	ch := make(chan ixBidResult)
-	for _, request := range requests {
-		go func(bidder *pbs.PBSBidder, request callOneObject) {
-			result, err := a.callOne(ctx, request.requestJSON)
-			result.Request = &request
-			result.Error = err
-			if result.Bid != nil {
-				result.Bid.BidderCode = bidder.BidderCode
-				result.Bid.BidID = bidder.LookupBidID(result.Bid.AdUnitCode)
-				result.Bid.Width = request.width
-				result.Bid.Height = request.height
-
-				if result.Bid.BidID == "" {
-					result.Error = &errortypes.BadServerResponse{
-						Message: fmt.Sprintf("Unknown ad unit code '%s'", result.Bid.AdUnitCode),
+		if request.Imp[0].Banner != nil {
+			banner := *request.Imp[0].Banner
+			request.Imp[0].Banner = &banner
+			formats := getBannerFormats(&banner)
+			for iFmt := range formats {
+				banner.Format = formats[iFmt : iFmt+1]
+				banner.W = openrtb.Uint64Ptr(banner.Format[0].W)
+				banner.H = openrtb.Uint64Ptr(banner.Format[0].H)
+				if requestData, err := createRequestData(a, request, &headers); err == nil {
+					if iFmt == 0 {
+						requests = append(requests, requestData)
+					} else {
+						multiSizeRequests = append(multiSizeRequests, requestData)
```

Comment:
Do you think keeping track of a multiSizeRequests counter would be easier than using two separate slices?

Reply:
I'm using two slices as a simple way to implement the logic described in the comments. If `createRequestData` could return an error it would be complicated to create one continuous slice in one loop.